cc: error: testsuite.o: No such file or directory

In attempting to set up the test suite, I'm hitting an error at [the dependencies step](https://i3wm.org/docs/testsuite.html#_installing_the_dependencies). `cpanm` seems to successfully install all dependencies of the `i3-testsuite` module, but fails at the `i3-testsuite` module itself:

```
~/work/extern/i3/testcases » cpanm .                                                                                
--> Working on .
Configuring /home/vladimir/work/extern/i3/testcases ... OK
Building and testing i3-testsuite- ... FAIL
! Installing . failed. See /home/vladimir/.cpanm/work/1504939586.23632/build.log for details. Retry with --force to force install it.
```

The referred log file contains:

```
cpanm (App::cpanminus) 1.7043 on perl 5.026000 built for x86_64-linux-thread-multi
Work directory is /home/vladimir/.cpanm/work/1504939586.23632
You have make /usr/sbin/make
You have LWP 6.26
You have /usr/sbin/tar: tar (GNU tar) 1.29
Copyright (C) 2015 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>.
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

Written by John Gilmore and Jay Fenlason.
You have /usr/sbin/unzip
--> Working on .
Entering /home/vladimir/work/extern/i3/testcases
Configuring /home/vladimir/work/extern/i3/testcases
Running Makefile.PL
Warning: NAME must be a package name
Generating a Unix-style Makefile
Writing Makefile for i3-testsuite
Writing MYMETA.yml and MYMETA.json
-> OK
Checking dependencies from MYMETA.json ...
Checking if you have AnyEvent 0 ... Yes (7.14)
Checking if you have ExtUtils::MakeMaker 0 ... Yes (7.24)
Checking if you have Test::More 0.94 ... Yes (1.302073)
Checking if you have IPC::Run 0 ... Yes (0.96)
Checking if you have X11::XCB 0.12 ... Yes (0.16)
Checking if you have ExtUtils::PkgConfig 0 ... Yes (1.16)
Checking if you have Inline 0 ... Yes (0.80)
Checking if you have Inline::C 0 ... Yes (0.78)
Building and testing i3-testsuite-
Running Mkbootstrap for testsuite ()
chmod 644 "testsuite.bs"
"/usr/bin/perl" -MExtUtils::Command::MM -e 'cp_nonempty' -- testsuite.bs blib/arch/auto/i3-testsuite/testsuite.bs 644
rm -f blib/arch/auto/i3-testsuite/testsuite.so
cc  -shared -Wl,-O1,--sort-common,--as-needed,-z,relro,-z,now -L/usr/local/lib -fstack-protector-strong testsuite.o  -o blib/arch/auto/i3-testsuite/testsuite.so  \
      \
  
cc: error: testsuite.o: No such file or directory
make: *** [Makefile:470: blib/arch/auto/i3-testsuite/testsuite.so] Error 1
-> FAIL Installing . failed. See /home/vladimir/.cpanm/work/1504939586.23632/build.log for details. Retry with --force to force install it.
```

What is this mysterious `testsuite.so` and what is it supposed to be made of?

I can think of one reason why things may have gone wrong, which is that I did not run `cpanm` via `sudo`, and instead set it up to install things to `~/perl5` according to the instructions it prints out, as I would like to avoid installing anything system-wide with anything other than my distribution's package manager.

The suggested `--force` flag seems to have no effect.

I found another mention of this error message in another issue #2849, though also without an explanation or solution.

I don't know the root cause, I get the same issue. I have worked around it the first time through trial and error. I `rm -rf`ed ~/.cpanm and ~/perl5 and this is how I got it working:

```
# From a clean i3 root:
$ autoreconf -fi && mkdir -p build && cd build
$ ../configure
$ make check -j # Fails with
Can't locate inc/Module/Install.pm in @INC

$ cpanm --local-lib=~/perl5 local::lib && eval $(perl -I ~/perl5/lib/perl5/ -Mlocal::lib)

$ cd ../testcases/
$ cpanm .
Building and testing X11-XCB-0.16 ... OK
Successfully installed X11-XCB-0.16
Building and testing i3-testsuite- ... FAIL
! Installing . failed. See /home/orestis/.cpanm/work/1505007528.23459/build.log for details. Retry with --force to force install it.
# build.log:
"/usr/bin/perl" -MExtUtils::Command::MM -e 'cp_nonempty' -- testsuite.bs blib/arch/auto/i3-testsuite/testsuite.bs 644
rm -f blib/arch/auto/i3-testsuite/testsuite.so
cc  -shared -Wl,-O1,--sort-common,--as-needed,-z,relro,-z,now -L/usr/local/lib -fstack-protector-strong testsuite.o  -o blib/arch/auto/i3-testsuite/testsuite.so  \
      \

cc: error: testsuite.o: No such file or directory
make: *** [Makefile:470: blib/arch/auto/i3-testsuite/testsuite.so] Error 1

$ cd ../AnyEvent-I3 # Let's check this one.
$ cpanm .
--> Working on .
Configuring /home/orestis/Documents/programming/i3/AnyEvent-I3 ... N/A
! Configure failed for .. See /home/orestis/.cpanm/work/1505008119.7481/build.log for details.
# build.log:
Can't locate inc/Module/Install.pm in @INC (you may need to install the inc::Module::Install module) (@INC contains: . /home/orestis/perl5/lib/perl5/5.26.0/x86_64-linux-thread-multi /home/orestis/perl5/lib/perl5/5.26.0 /home/orestis/perl5/lib/perl5/x86_64-linux-thread-multi /home/orestis/perl5/lib/perl5 /usr/lib/perl5/5.26/site_perl /usr/share/perl5/site_perl /usr/lib/perl5/5.26/vendor_perl /usr/share/perl5/vendor_perl /usr/lib/perl5/5.26/core_perl /usr/share/perl5/core_perl) at Makefile.PL line 2.
BEGIN failed--compilation aborted at Makefile.PL line 2.
-> N/A
-> FAIL Configure failed for .. See /home/orestis/.cpanm/work/1505008119.7481/build.log for details.

$ cpanm Module::Install
$ cpanm .
$ cd ../testcases/
$ cpanm . # Fails again.
$ cd ..
$ git clean -fdx . # This is probably important for some reason.
$ autoreconf -fi && mkdir -p build && cd build
$ ../configure
$ make check -j # Tests now work (?!)
```

--- a/testcases/Makefile.PL
+++ b/testcases/Makefile.PL
@@ -18,7 +18,11 @@ WriteMakefile(
     PM => {}, # do not install any files from this directory
     clean => {
         FILES => 'testsuite-* latest i3-cfg-for-*',
-    }
+    },
+    # This is a pure-Perl distribution:
+    linkext => {
+        LINKTYPE => '',
+    },
 );
 
 package MY;

